pysb/tools/sensitivity_analysis.py

```diff
@@ -0,0 +1,359 @@
+        self.values_to_sample = values_to_sample
+        self.n_sam = len(self.values_to_sample)
+        self.b_matrix = []
+        self.b_prime_matrix = []
+        self.nm = self.n_proteins * self.n_sam
+        self.size_of_matrix = self.nm ** 2
+        self.shape_of_matrix = (self.nm, self.nm)
+        self.initial_conditions = np.zeros(len(self.model.species))
+        self.index_of_species_of_interest = self.create_index_of_species()
+        self.setup_sampling_matrix()
+        self.simulations = self._find_redundant_simulations()
+        self.objective_function = objective_function
+        self.observable = observable
+        self.standard = None
+        self.p_prime_matrix = np.zeros(self.size_of_matrix)
+        self.p_matrix = np.zeros(self.size_of_matrix)
+
+    def _calculate_objective(self, species):
+        """ Calculates fraction of change between standard value and newly obtained value
+
+        :param species:
+        :return: fraction of change, type float
+        """
+        return (self.objective_function(
+                species) - self.standard) / self.standard * 100.
+
+    def run(self, run_solver, save_name, out_dir):
+        """
+
+        :param run_solver:
+        :param save_name:
+        :param out_dir:
+        """
+        if not os.path.exists(out_dir):
+            os.mkdir(out_dir)
+        solver = pysb.integrate.Solver(self.model, self.sim_time, rtol=1e-8,
+                                       atol=1e-8, integrator='lsoda',
+                                       mxstep=20000)
+        solver.run()
+        self.standard = self.objective_function(solver.yobs[self.observable])
+        sensitivity_output = run_solver(self.simulations)
+
+        p_matrix = np.zeros(self.size_of_matrix)
+        p_prime_matrix = np.zeros(self.size_of_matrix)
+        counter = 0
+        for i in range(len(p_matrix)):
+            if i in self.b_index:
+                tmp = self._calculate_objective(sensitivity_output[:, counter])
+                p_matrix[i] = tmp
+                counter += 1
+
+        for i in range(len(p_matrix)):
+            if i in self.b_prime_not_in_b:
+                tmp = self._calculate_objective(
+                        sensitivity_output[:, self.b_prime_not_in_b[i]])
+                p_prime_matrix[i] = tmp
+            elif i in self.b_prime_in_b:
+                p_prime_matrix[i] = p_matrix[self.b_prime_in_b[i]]
+        # Reshape
+        p_matrix = p_matrix.reshape(self.shape_of_matrix)
+        # Project the mirrored image
+        self.p_matrix = p_matrix + p_matrix.T
+        self.p_prime_matrix = p_prime_matrix.reshape(self.shape_of_matrix)
+
+        np.savetxt(os.path.join(out_dir, '%s_p_matrix.csv' % save_name),
+                   self.p_matrix)
+        np.savetxt(os.path.join(out_dir, '%s_p_prime_matrix.csv' % save_name),
+                   self.p_prime_matrix)
+
+    def create_index_of_species(self):
```

Comment:
Methods like this should probably be private if the end user doesn't need to call them directly (i.e. name should have a leading underscore)